tritium-processor/src/main/java/com/palantir/tritium/processor/TritiumAnnotationProcessorStrategy.java

```diff
@@ -0,0 +1,282 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.processor;
+
+import com.palantir.delegate.processors.AnnotatedType;
+import com.palantir.delegate.processors.AnnotatedTypeMethod;
+import com.palantir.delegate.processors.DelegateProcessorStrategy;
+import com.palantir.delegate.processors.LocalVariable;
+import com.palantir.tritium.annotations.Instrument;
+import com.palantir.tritium.annotations.internal.InstrumentationBuilder;
+import com.palantir.tritium.api.event.InstrumentationFilter;
+import com.palantir.tritium.event.Handlers;
+import com.palantir.tritium.event.InvocationContext;
+import com.palantir.tritium.event.InvocationEventHandler;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.CodeBlock;
+import com.squareup.javapoet.FieldSpec;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.ParameterSpec;
+import com.squareup.javapoet.ParameterizedTypeName;
+import com.squareup.javapoet.TypeName;
+import com.squareup.javapoet.TypeSpec;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeMirror;
+import javax.tools.Diagnostic.Kind;
+
+enum TritiumAnnotationProcessorStrategy implements DelegateProcessorStrategy {
+    INSTANCE;
+
+    private static final String FILTER_NAME = "filter";
+    private static final String HANDLER_NAME = "handler";
+    private static final String CONTEXT_NAME = "_invocationContext";
+
+    @Override
+    public Set<String> supportedAnnotations() {
+        return Set.of(Instrument.class.getName());
+    }
+
+    @Override
+    public String generatedTypeName(String annotatedTypeName) {
+        return "Instrumented" + annotatedTypeName;
+    }
+
+    @Override
+    public TypeName delegateType(DelegateTypeArguments arguments) {
+        TypeElement type = arguments.type().type();
+        TypeName typeName = TypeName.get(type.asType());
+
+        if (type.getKind() != ElementKind.INTERFACE) {
+            arguments
+                    .context()
+                    .messager()
+                    .printMessage(
+                            Kind.ERROR,
+                            "Only interfaces may be instrumented using @" + Instrument.class.getSimpleName(),
+                            arguments.type().type());
+            return typeName;
+        }
```

Comment:
I suppose we could relax this constraint in the future, allowing `@Instrument` on concrete implementations to generate a wrapper based on its interfaces -- for a separate PR if you think it would be useful.

Reply:
Yeah I played around with this a bit, but it ended up being a bit tricky. Particularly because we need a return type for the static factory methods. Currently [we're using the annotated type](https://github.com/palantir/tritium/blob/c9d266d82b7cadf86e6520060dd9ff70afaead7a/tritium-processor/src/main/java/com/palantir/tritium/processor/TritiumAnnotationProcessorStrategy.java#L251), but this is incorrect if the annotated type is a class - you'll get something like:

```java
public final class InstrumentedAnnotatedClass implements Supplier<String> {
    ...

    public static AnnotatedClass instrument(AnnotatedClass delegate, TaggedMetricRegistry registry) {
            return builder(delegate).withTaggedMetrics(registry).withTracing().build();
    }
}
```

You might be able to make this work when there is a single interface, but it's much more complicated if the class implements multiple interfaces. This is partly why I'm not convinced that delegate processors should even support this case. This would also change the behavior when instrumenting an external interface - the `InstrumentedDelegateToCallable` factory methods would return `Callable` instead of `DelegateToCallable`.

I think there's value in limiting instrumentation to only interface, even if it's technically possible to support classes.